src/cargo/core/resolver/mod.rs

```diff
@@ -256,12 +257,43 @@ impl<'a> Iterator for DepsNotReplaced<'a> {
     }
 }
 
+enum RcList<T> {
+    Data(Rc<(T, RcList<T>)>),
+    Empty,
+}
+
+impl<T> RcList<T> {
+    fn new() -> RcList<T> {
+        RcList::Empty
+    }
+
+    fn push(&mut self, data: T) {
+        let node = Rc::new((data, mem::replace(self, RcList::Empty)));
+        *self = RcList::Data(node);
+    }
+}
+
+// Not derived to avoid `T: Clone`
+impl<T> Clone for RcList<T> {
+    fn clone(&self) -> RcList<T> {
+        match *self {
+            RcList::Data(ref d) => RcList::Data(d.clone()),
+            RcList::Empty => RcList::Empty,
+        }
+    }
+}
+
+enum GraphNode {
+    Add(PackageId),
+    Link(PackageId, PackageId),
+}
+
 #[derive(Clone)]
 struct Context<'a> {
-    activations: HashMap<(String, SourceId), Vec<Rc<Summary>>>,
-    resolve_graph: Graph<PackageId>,
+    activations: HashMap<String, HashMap<SourceId, Vec<Summary>>>,
```

Comment:
Yeah this is something I need to add a `TODO` for. The fields of `activations` and the binary heap of remaining deps are still cloned (if needed) on all activates. These are the number one and two costs in resolution. They just disappeared in this PR b/c of the fast path which avoids cloning.

I thought for awhile about how to work with this but yeah I think the nested map being some sort of "small" thing makes sense. I was hoping we could hold out and wait for a persistent hash map though to avoid too many crazy data structures.

Once I saw the "don't clone as much" optimization erase this from the profile though I figured it'd be best to move on to the rest of Cargo :). Right now loading info from the index and parsing TOML is pretty high up there. Those I think would ideally be tackled next.